src/ctxt.ml

```diff
@@ -3,37 +3,64 @@
 open Extra
 open Terms
 
-(** Representation of a typing context, associating a type (or [Term.term]) to
-    free [Bindlib] variables. *)
-type t = (tvar * term) list
+type t = ctxt
 
 (** [empty] is the empty context. *)
 let empty : t = []
 
-(** [add x a ctx] maps the variable [x] to the type [a] in [ctx]. *)
-let add : tvar -> term -> t -> t =
-  fun x a ctx -> (x,a)::ctx
+(** [assume h ctx] adds list of assumptions [h] mapping variables to their
+    types to context [ctx]. *)
+let assume : (tvar * term) list -> t -> t = fun hyps ctx ->
+  List.map (fun (x,a) -> (x, a, None)) hyps @ ctx
 
-(** [unbind_ctxt ctx a b] returns the triple [(x,b,ctx')] such that [(x,b)]
-   is the unbinding of [b] and [ctx'] is the context [ctx] extended with
-   [(x,a)] if [x] occurs in [b]. *)
-let unbind ctx a b =
+(** [define x a t ctx] maps variable [x] to term [t] of type [a] in [ctx]. *)
+let define : tvar -> term -> term -> t -> t = fun x a t ctx ->
+  (x, a, Some(t)) :: ctx
+
+(** [unbind_ctxt ctx a ?def b] returns the triple [(x,b,ctx')] such that
+    [(x,b)] is the unbinding of [b] and [ctx'] is the context [ctx] extended
+    with, if [x] occurs in [b]
+    - [Assume(x, a)] if [?def] is not given and
+    - [Define{ctx_v=x; ctx_y=a; ctx_e=?def}] otherwise. *)
+let unbind : t -> term -> ?def:term -> tbinder -> tvar * term * t =
+  fun ctx a ?def b ->
   let (x,b') = Bindlib.unbind b in
-  let ctx' = if Bindlib.binder_occur b then add x a ctx else ctx in
+  let ctx' =
+    if Bindlib.binder_occur b then
+      match def with
+      | None    -> assume [(x,a)] ctx
+      | Some(t) -> define x a t ctx
+    else ctx
+  in
   (x,b',ctx')
 
-(** [pp oc ctx] prints the context [ctx] to the channel [oc]. *)
-let pp : t pp = fun oc ctx ->
-  let pp_e oc (x,a) =
-    Format.fprintf oc "%a : %a" Print.pp_tvar x Print.pp a
-  in
-  if ctx = [] then Format.pp_print_string oc "∅"
-  else List.pp pp_e ", " oc (List.rev ctx)
+(** [assumptions ctx] returns a list mapping free variable to their types. *)
+let assumptions : t -> (tvar * term) list =
```

Comment:
should probably be removed once `(tvar*term) list` is replaced by `ctxt`

Reply:
same as for `assume` we can decide as well to not abstract anything and just throw away this function.

Reply:
I'll remove all this abstraction thing, it is simpler without.

Reply:
`ctxt` is not `(tvar * term) list`!

Reply:
I know. But I think that every `(tvar*term)list` should be replaced by `ctxt` for simplicity.